### PR TITLE
[ISSUE-1124] - Windows Bolt support

### PR DESCRIPTION
**Pull request by jerrypeng**
https://github.com/twitter/heron/issues/1124

Most of the windowing code comes from the Apache Storm

Cherry picked most of the code from the master branch of Apache Storm

### CODE REVIEW

**Comment by jerrypeng** [comment]
@srkukarni Thanks for the review.  I have addressed your comment

File: heron/api/src/java/com/twitter/heron/api/bolt/WindowedBoltExecutor.java [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+    this.windowManager = initWindowManager(listener, topoConf, context, queue);
+    start();
+    LOG.info(String.format("Initialized window manager %s", windowManager));
+  }
+
+  @Override
+  public void execute(Tuple input) {
+    if (isTupleTs()) {
+      long ts = timestampExtractor.extractTimestamp(input);
+      if (waterMarkEventGenerator.track(input.getSourceGlobalStreamId(), ts)) {
```
Comment (srkukarni) [inline heron/api/src/java/com/twitter/heron/api/bolt/WindowedBoltExecutor.java]: Could it be possible not to use getSourceGlobalStreamId(). One of the reasons that we did not implement this in heron was that we felt it exposed too low level details. Could we instead get the SourceComponent and SourceStreamId as individual parameters?

File: heron/api/src/java/com/twitter/heron/api/windowing/EvictionPolicy.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package com.twitter.heron.api.windowing;
+
+/**
+ * Eviction policy tracks events and decides whether
+ * an event should be evicted from the window or not.
+ *
+ * @param <T> the type of event that is tracked.
```
Comment (srkukarni) [inline heron/api/src/java/com/twitter/heron/api/windowing/EvictionPolicy.java]: What does S stand for?

File: heron/api/src/java/com/twitter/heron/api/bolt/WindowedBoltExecutor.java (hunk 1 above)
Comment (jerrypeng) [reply heron/api/src/java/com/twitter/heron/api/bolt/WindowedBoltExecutor.java]: @srkukarni I understand what you are saying, but there are already places in the Heron Java API that use proto buf objects e.g. TopologyContext:

Map<TopologyAPI.StreamId, TopologyAPI.Grouping> getThisSources()

File: heron/api/src/java/com/twitter/heron/api/windowing/EvictionPolicy.java (hunk 2 above)
Comment (jerrypeng) [reply heron/api/src/java/com/twitter/heron/api/windowing/EvictionPolicy.java]: will remove.  This had to do with state implementation along side windowing, but we will not pulling in that piece yet

File: heron/api/src/java/com/twitter/heron/api/windowing/DefaultEvictionContext.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+package com.twitter.heron.api.windowing;
+
+public class DefaultEvictionContext implements EvictionContext {
+  private final Long referenceTime;
```
[the remark this answers is not in the document]
Comment (jerrypeng) [reply heron/api/src/java/com/twitter/heron/api/windowing/DefaultEvictionContext.java]: will add

**Comment by jerrypeng** [comment]
@srkukarni I have address your comments

File: heron/api/tests/java/BUILD [hunk 4]
```diff
@@ -0,0 +1,26 @@
+load("/tools/rules/java_tests", "java_tests")
+load("/tools/rules/heron_deps", "heron_java_proto_files")
+
+api_deps_files = [
+    "//heron/api/src/java:api-java",
+    "//third_party/java:junit4",
+]
+
+api_deps_files = \
+  heron_java_proto_files() + \
```
Comment (srkukarni) [inline heron/api/tests/java/BUILD]: can we have heron_java_api_proto_files instead

Comment (jerrypeng) [reply heron/api/tests/java/BUILD]: sure

**Comment by jerrypeng** [comment]
@srkukarni thanks for the review.  I think I have addressed your last round of coments

**Comment by srkukarni** [comment]
looks good!

**Pull request merged**